File: 152-maximum-product-subarray/maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        // normally if a product is less than zero store it in diff var 
      int max_product, ans , min_product;
      ans =max_product=min_product=nums[0];
      for(int i =1;i<nums.size();i++){
        int temp = max_product;
        max_product = max ({nums[i],max_product*nums[i], min_product*nums[i]});
        min_product = min ({nums[i],nums[i]* temp, nums[i]*min_product});
        ans = max(ans,max_product);
      }
      return ans ;

    }
};

File: 152-maximum-product-subarray/maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let ans = nums[0];
    let max_product =nums[0];
    let min_product=nums[0];
    for(let i=1;i<nums.length;i++){
       let temp=max_product;
       max_product=Math.max(nums[i],max_product*nums[i],min_product*nums[i]);
       min_product=Math.min(nums[i],temp*nums[i],min_product*nums[i]);
       ans=Math.max(ans,max_product)

    
    }
    return ans;
};